File: recruitpro-api/app/services/notification_service.py
from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.repositories.notification_repository import NotificationRepository
from app.schemas.notification_schema import NotificationCreate
from app.models.user import User


class NotificationService:

    def __init__(self):
        self.repository = NotificationRepository()

    # ==================================================
    # GET COMPANY ID
    # ==================================================
    def get_company_id(self, current_user: dict):
        company_id = current_user.get("company_id")

        if not company_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Company information not found for the current user",
            )

        return company_id

    # ==================================================
    # GET USER ID
    # ==================================================
    def get_user_id(self, current_user: dict):
        user_id = current_user.get("user_id")

        if not user_id:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User information not found",
            )

        return user_id

    # ==================================================
    # VALIDATE RECEIVER USER
    # ==================================================
    def validate_receiver(
        self,
        db: Session,
        user_id: int,
        company_id: int,
    ):
        user = (
            db.query(User)
            .filter(
                User.UserId == user_id,
                User.CompanyId == company_id,
            )
            .first()
        )

        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Notification receiver not found in your company",
            )

        return user

    # ==================================================
    # GET ALL NOTIFICATIONS
    # ==================================================
    def get_all_notifications(
        self,
        db: Session,
        current_user: dict,
        search: str = "",
        is_read: bool | None = None,
        notification_type: str | None = None,
        sort_by: str = "CreatedOn",
        order: str = "desc",
        page: int = 1,
        page_size: int = 10,
    ):
        company_id = self.get_company_id(current_user)
        user_id = self.get_user_id(current_user)

        return self.repository.get_all(
            db=db,
            company_id=company_id,
            user_id=user_id,
            search=search,
            is_read=is_read,
            notification_type=notification_type,
            sort_by=sort_by,
            order=order,
            page=page,
            page_size=page_size,
        )

    # ==================================================
    # GET NOTIFICATION BY ID
    # ==================================================
    def get_notification_by_id(
        self,
        db: Session,
        notification_id: int,
        current_user: dict,
    ):
        company_id = self.get_company_id(current_user)
        user_id = self.get_user_id(current_user)

        notification = self.repository.get_by_id(
            db=db,
            notification_id=notification_id,
            company_id=company_id,
            user_id=user_id,
        )

        if not notification:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Notification not found",
            )

        return notification

    # ==================================================
    # GET UNREAD COUNT
    # ==================================================
    def get_unread_count(
        self,
        db: Session,
        current_user: dict,
    ):
        company_id = self.get_company_id(current_user)
        user_id = self.get_user_id(current_user)

        unread_count = self.repository.get_unread_count(
            db=db,
            company_id=company_id,
            user_id=user_id,
        )

        return {
            "unread_count": unread_count,
        }

    # ==================================================
    # CREATE NOTIFICATION
    # ==================================================
    def create_notification(
        self,
        db: Session,
        notification: NotificationCreate,
        current_user: dict,
    ):
        company_id = self.get_company_id(current_user)

        # ==================================================
        # VALIDATE RECEIVER
        # ==================================================
        self.validate_receiver(
            db=db,
            user_id=notification.UserId,
            company_id=company_id,
        )

        # ==================================================
        # CREATE NOTIFICATION
        # ==================================================
        return self.repository.create(
            db=db,
            notification=notification,
            company_id=company_id,
        )

    # ==================================================
    # MARK ONE NOTIFICATION AS READ
    # ==================================================
    def mark_as_read(
        self,
        db: Session,
        notification_id: int,
        current_user: dict,
    ):
        company_id = self.get_company_id(current_user)
        user_id = self.get_user_id(current_user)

        notification = self.repository.mark_as_read(
            db=db,
            notification_id=notification_id,
            company_id=company_id,
            user_id=user_id,
        )

        if not notification:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Notification not found",
            )

        return {
            "message": "Notification marked as read successfully",
            "NotificationId": notification.NotificationId,
            "IsRead": notification.IsRead,
        }

    # ==================================================
    # MARK ALL NOTIFICATIONS AS READ
    # ==================================================
    def mark_all_as_read(
        self,
        db: Session,
        current_user: dict,
    ):
        company_id = self.get_company_id(current_user)
        user_id = self.get_user_id(current_user)

        updated_count = self.repository.mark_all_as_read(
            db=db,
            company_id=company_id,
            user_id=user_id,
        )

        return {
            "message": "All notifications marked as read successfully",
            "updated_count": updated_count,
        }

File: recruitpro-ui/src/app/services/notification.service.ts
import { Injectable } from '@angular/core';
import {
    HttpClient,
    HttpParams
} from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
    providedIn: 'root'
})
export class NotificationService {

    // ==================================================
    // API URL
    // ==================================================
    private apiUrl =
        'http://127.0.0.1:8000/notifications';

    constructor(
        private http: HttpClient
    ) { }

    // ==================================================
    // GET ALL NOTIFICATIONS
    // ==================================================
    getNotifications(
        search: string = '',
        isRead: boolean | null = null,
        notificationType: string | null = null,
        sortBy: string = 'CreatedOn',
        order: string = 'desc',
        page: number = 1,
        pageSize: number = 10
    ): Observable<any> {

        let params = new HttpParams()
            .set('search', search)
            .set('sort_by', sortBy)
            .set('order', order)
            .set('page', page.toString())
            .set('page_size', pageSize.toString());

        // ==================================================
        // READ / UNREAD FILTER
        // ==================================================
        if (isRead !== null) {
            params = params.set(
                'is_read',
                isRead.toString()
            );
        }

        // ==================================================
        // NOTIFICATION TYPE FILTER
        // ==================================================
        if (
            notificationType &&
            notificationType !== 'All'
        ) {
            params = params.set(
                'notification_type',
                notificationType
            );
        }

        return this.http.get<any>(
            `${this.apiUrl}/`,
            { params }
        );
    }

    // ==================================================
    // GET UNREAD COUNT
    // ==================================================
    getUnreadCount(): Observable<any> {

        return this.http.get<any>(
            `${this.apiUrl}/unread-count`
        );
    }

    // ==================================================
    // GET NOTIFICATION BY ID
    // ==================================================
    getNotificationById(
        notificationId: number
    ): Observable<any> {

        return this.http.get<any>(
            `${this.apiUrl}/${notificationId}`
        );
    }

    // ==================================================
    // ADD NOTIFICATION
    // ==================================================
    addNotification(
        data: any
    ): Observable<any> {

        return this.http.post<any>(
            `${this.apiUrl}/`,
            data
        );
    }

    // ==================================================
    // MARK NOTIFICATION AS READ
    // ==================================================
    markAsRead(
        notificationId: number
    ): Observable<any> {

        return this.http.put<any>(
            `${this.apiUrl}/${notificationId}/read`,
            {}
        );
    }

    // ==================================================
    // MARK ALL NOTIFICATIONS AS READ
    // ==================================================
    markAllAsRead(): Observable<any> {

        return this.http.put<any>(
            `${this.apiUrl}/mark-all-read`,
            {}
        );
    }
}